theano/gpuarray/c_code/dnn_sptf.c

```diff
@@ -0,0 +1,203 @@
+                         cudnnSpatialTransformerDescriptor_t desc,
+                         PyGpuArrayObject ** output,
+                         PyGpuArrayObject ** grid,
+                         cudnnHandle_t _handle)
+{
+    PyGpuContextObject * gpu_ctx = input->context;
+    void * alpha_p;
+    void * beta_p;
+    double alpha = 1.0, beta = 0.0;
+    float af = alpha, bf = beta;
+    cudnnStatus_t err = CUDNN_STATUS_SUCCESS;
+    int num_images, num_channels, height, width;
+    size_t gpu_grid_dims[4], out_dims[4];
+
+    switch (input->ga.typecode)
+    {
+    case GA_DOUBLE:
+        alpha_p = (void *)&alpha;
+        beta_p = (void *)&beta;
+        break;
+    case GA_FLOAT:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        break;
+    case GA_HALF:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        break;
+    default:
+        PyErr_SetString( PyExc_TypeError,
+            "GpuDnnTransformer: unsupported type for input in spatial transformer." );
+        return 1;
+    }
+
+    if ( theta->ga.typecode != GA_FLOAT &&
+         theta->ga.typecode != GA_DOUBLE &&
+         theta->ga.typecode != GA_HALF )
+    {
+        PyErr_SetString( PyExc_TypeError,
+            "GpuDnnTransformer: unsupported data type for theta in spatial transformer." );
+        return 1;
+    }
+    else if ( PyGpuArray_DIM( theta, 1 ) != 2 && PyGpuArray_DIM( theta, 2 ) != 3 )
+    {
+        PyErr_Format( PyExc_RuntimeError,
+            "GpuDnnTransformer: incorrect dimensions for theta, expected (%d, %d, %d), got (%d, %d, %d)",
+            PyGpuArray_DIMS( theta )[0], 2, 3, PyGpuArray_DIMS( theta )[0],
+            PyGpuArray_DIMS( theta )[1], PyGpuArray_DIMS( theta )[2] );
+        return 1;
+    }
+
+    if ( PyArray_NDIM( grid_dims ) != 1 || PyArray_SIZE( grid_dims ) != 4 )
+    {
+        PyErr_SetString( PyExc_MemoryError,
+            "GpuDnnTransformer: grid_dims must have 4 elements." );
+        return 1;
+    }
+
+    // Obtain grid dimensions
+    num_images = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 0 ) );
+    num_channels = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 1 ) );
+    height = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 2 ) );
+    width = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 3 ) );
+
+    gpu_grid_dims[0] = num_images;
+    gpu_grid_dims[1] = height;
+    gpu_grid_dims[2] = width;
+    gpu_grid_dims[3] = 2;
+
+    out_dims[0] = num_images;
+    out_dims[1] = num_channels;
+    out_dims[2] = height;
+    out_dims[3] = width;
+
+    if ( width == 0 || height == 0 || num_images == 0 )
+    {
+        PyErr_SetString( PyExc_RuntimeError,
+            "GpuDnnTransformer: grid_dims has a dimension with value zero" );
+        return 1;
+    }
+
+    if ( PyGpuArray_DIM( input, 0 ) != num_images )
+    {
+        PyErr_Format( PyExc_RuntimeError,
+            "GpuDnnTransformer: expected input to have %d inputs, got %d inputs.",
```

Comment:
Consider rewrite this message, e.g.: "GpuDnnTransformer: expected batch size %d, got %d." .